registry/client_factory.go

```diff
@@ -98,9 +123,14 @@ func (f *RemoteClientFactory) ClientFor(repo image.CanonicalName, creds Credenti
 		f.Logger.Log("repo", repo.String(), "auth", cred.String(), "api", registryURL.String())
 	}
 
-	tokenHandler := auth.NewTokenHandler(tx, &store{cred}, repo.Image, "pull")
-	basicauthHandler := auth.NewBasicHandler(&store{cred})
-	tx = transport.NewTransport(tx, auth.NewAuthorizer(manager, tokenHandler, basicauthHandler))
+	authHandlers := []auth.AuthenticationHandler{}
+	// only send creds over HTTPS
+	if registryURL.Scheme == "https" {
```

Comment:
I think this is too restrictive and opinionated. For instance, one may intentionally want to use http in a sealed environment in which the registry just happens to require credentials for reasons beyond your control.

Also, it is almost equally unsafe to send the creds over HTTPS when the certificate verification is being skipped

Reply:
Yeah those two points are fair. So I guess the recommendation would be to

 1. only fall back to http if the domain is flagged as insecure
 2. only use TLS_INSECURE_SKIP_VERIFY if the domain is flagged as insecure
 3. always use credentials if you have them

on the basis that cases 1. and 2. are basically equivalent, and if you've flagged the domain as insecure, you are accepting the risk of a MitM attack in either situation -- ?

Reply:
Yep, that sounds good

Reply:
^^ @Timer @donifer @cazzoo Just to bring this significant change to your attention, since you've indicated this PR may be of interest to you